Cabal/Distribution/Simple/Build.hs

```diff
@@ -586,5 +592,15 @@ writeAutogenFiles verbosity pkg lbi clbi = do
                  </> ModuleName.toFilePath (autogenPathsModuleName pkg) <.> "hs"
   rewriteFile pathsModulePath (Build.PathsModule.generate pkg lbi clbi)
 
+  case clbi of
+    LibComponentLocalBuildInfo { componentInstantiatedWith = insts } ->
+        -- Harmless enough to do things even when they exist
+        for_ (map fst insts) $ \mod_name -> do
+            let sigPath = autogenComponentModulesDir lbi clbi
+                      </> ModuleName.toFilePath mod_name <.> "hsig"
+            createDirectoryIfMissingVerbose verbosity True (takeDirectory sigPath)
+            rewriteFile sigPath $ "signature " ++ display mod_name ++ " where"
+    _ -> return ()
+
```

Comment:
Would be nice to get a comment about what this new section is doing, as it's not obvious.